models/cross-spork_client.go

```diff
@@ -0,0 +1,124 @@
+package models
+
+import (
+	"context"
+	"fmt"
+	"github.com/onflow/cadence"
+	"github.com/onflow/flow-go-sdk"
+	"github.com/onflow/flow-go-sdk/access"
+	"github.com/onflow/flow-go-sdk/access/grpc"
+	"github.com/rs/zerolog"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+)
+
+// CrossSporkClient is a wrapper around the Flow AN client that can
+// access different AN APIs based on the height boundaries of the sporks.
+//
+// Each spork is defined with the last height included in that spork,
+// based on the list we know which AN client to use when requesting the data.
+//
+// Any API that supports cross-spork access must have a defined function
+// that shadows the original access Client function.
+type CrossSporkClient struct {
+	logger zerolog.Logger
+	// this map holds the last heights and clients for each spork
+	sporkHosts map[uint64]access.Client
+
+	access.Client
+}
+
+// NewCrossSporkClient creates a new instance of the client, it accepts the
+// host to the current spork AN API.
+func NewCrossSporkClient(currentSporkHost string, logger zerolog.Logger) (*CrossSporkClient, error) {
+	// add current spork AN host as the default client
+	client, err := grpc.NewClient(currentSporkHost)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CrossSporkClient{
+		logger,
+		make(map[uint64]access.Client),
+		client,
+	}, nil
+}
+
+func (c *CrossSporkClient) AddSpork(lastHeight uint64, host string) error {
+	if _, ok := c.sporkHosts[lastHeight]; ok {
+		return fmt.Errorf("provided last height already exists")
+	}
+
+	client, err := grpc.NewClient(host)
+	if err != nil {
+		return err
+	}
+
+	c.sporkHosts[lastHeight] = client
```

Comment:
it's a good idea to validate the range to make sure there is a contiguous range of blocks. You can use `GetNodeVersionInfo()` to get the first block (`NodeRootBlockHeight` is the first block the node has):
https://github.com/onflow/flow-go/blob/1d581d472b7dcc40cd8952da497422c910ebd99b/access/api.go#L259-L266

`GetLatestBlockHeader()` to get the latest sealed block.

Then you can search all nodes to check for gaps

Reply:
That's a good idea, didn't know it exists. In my mind if you wrongly configured it would err out when accessing, but this is better for sure.

Reply:
I don't have access to this method through the SDK

Reply:
I can issue this up as a follow up PR since I believe it will require updating Go SDK first and then updating here.

Reply:
https://github.com/onflow/flow-evm-gateway/issues/231